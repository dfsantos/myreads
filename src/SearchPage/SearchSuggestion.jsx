import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'underscore';

const propTypes = {
  words: PropTypes.arrayOf(PropTypes.string).isRequired,
  pause: PropTypes.bool,
};

const defaultProps = {
  pause: false,
};

class SearchSuggestion extends Component {
  constructor(props) {
    super(props);
    this.state = { words: _.shuffle(props.words) };
    this.suggestWord = this.suggestWord.bind(this);
    this.stop = setInterval(this.suggestWord, 2500);
  }

  componentWillUnmount() {
    clearInterval(this.stop);
  }

  suggestWord() {
    this.setState(state => {
      const { words } = state;
      words.unshift(words.pop());
      return { words };
    });
  }

  render() {
    return (
      <div>
        {!this.props.pause && (
          <div className="search-suggestion">
            Try search for <br />
            "{this.state.words[0]}"
          </div>
        )}
      </div>
    );
  }
}

SearchSuggestion.propTypes = propTypes;
SearchSuggestion.defaultProps = defaultProps;

export default SearchSuggestion;
