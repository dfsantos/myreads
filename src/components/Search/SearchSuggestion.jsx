import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'underscore';

const propTypes = {
  words: PropTypes.arrayOf(PropTypes.string).isRequired,
  pause: PropTypes.bool,
};

const defaultProps = {
  pause: false,
};

class SearchSuggestion extends Component {
  constructor(props) {
    super(props);
    this.state = { words: _.shuffle(props.words) };
    this.suggestWord = this.suggestWord.bind(this);
    this.stop = setInterval(this.suggestWord, 2500);
  }

  componentWillReceiveProps(nextProps) {
    if (!nextProps.pause && this.stop === null) {
      this.stop = setInterval(this.suggestWord, 2500);
    } else {
      clearInterval(this.stop);
      this.stop = null;
    }
  }

  componentWillUnmount() {
    clearInterval(this.stop);
  }

  suggestWord() {
    this.setState(state => {
      const { words } = state;
      words.unshift(words.pop());
      return { words };
    });
  }

  render() {
    const { pause } = this.props;
    const { words } = this.state;
    return (
      <div>
        {!pause && (
          <div className="search-suggestion">Try search for "{words[0].toLowerCase()}"</div>
        )}
      </div>
    );
  }
}

SearchSuggestion.propTypes = propTypes;
SearchSuggestion.defaultProps = defaultProps;

export default SearchSuggestion;
