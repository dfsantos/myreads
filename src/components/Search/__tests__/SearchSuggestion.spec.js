import React from 'react';

import SearchSuggestion from '../SearchSuggestion';

const words = ['Android'];

it('it should render without errors', () => {
  shallow(<SearchSuggestion words={words} />);
});

it('snapshot', () => {
  const wrapper = shallow(<SearchSuggestion words={words} />);
  expect(toJson(wrapper)).toMatchSnapshot();
});

it('setInterval should be called with interval ID when componentWillReceiveProps method execute', () => {
  const wrapper = shallow(<SearchSuggestion words={words} />);
  wrapper.instance().stop = 1;
  wrapper.setProps({ pause: true });
  expect(setInterval).toHaveBeenCalled();
  expect(wrapper.instance().stop).toBeNull();
});

it('setInterval should be activate when componentWillReceiveProps method execute and this.stop is null', () => {
  const wrapper = shallow(<SearchSuggestion words={words} />);
  wrapper.instance().stop = null;
  wrapper.setProps({ pause: false });
  expect(setInterval).toHaveBeenCalledWith(wrapper.instance().suggestWord, 2500);
});

it('clearInterval should be called when componentWillUnmount method execute', () => {
  const wrapper = shallow(<SearchSuggestion words={words} />);
  wrapper.unmount();
  expect(clearInterval).toHaveBeenCalled();
});
