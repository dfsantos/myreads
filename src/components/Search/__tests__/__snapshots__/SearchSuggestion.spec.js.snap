// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <div
    className="search-suggestion"
  >
    Try search for "
    java
    "
  </div>
</div>
`;
